[CORL-1016] Admin Dashboard

tessalt opened the pull request:
<!--

Thank you for submitting a pull request! Please note that by contributing to
Coral, you agree to our Code of Conduct: http://code-of-conduct.voxmedia.com/

Before submitting your Pull Request (or PR), please verify that:

* [ ] Your code is up-to-date with the base branch
* [ ] You've successfully run `npm run test` locally

Refer to CONTRIBUTING.MD for more details.

  https://github.com/coralproject/talk/blob/master/CONTRIBUTING.md

-->

## What does this PR do?

<!--

In this section, you should be describing what other Github issues or tickets
that this PR is designed to addressed.

Any related Github issue should be linked by adding its URL to this section.

-->

## What changes to the GraphQL/Database Schema does this PR introduce?

<!--

In this section, you should describe any changes to be made to the GraphQL
schema file (located https://github.com/coralproject/talk/blob/master/src/core/server/graph/schema/schema.graphql) or any
database model (located as types in the https://github.com/coralproject/talk/blob/master/src/core/server/models directory).

If no changes were added to the GraphQL/Database Schema as a part of this PR,
simply write "None".

-->

## How do I test this PR?

<!--

In this section, you should be describing any manual testing that can be used to
verify features introduced or bugs fixed in this PR.

 -->

tessalt committed:
count things in redis

tessalt committed:
add rest routes for daily comment totals

tessalt committed:
add route for new commenters

tessalt committed:
retrieve hourly totals for new commenters and comments

tessalt committed:
add rate limiting

tessalt committed:
add routes and components for dashboard

tessalt committed:
add top stories component

tessalt committed:
add activity charts

tessalt committed:
clean up rest responses

tessalt committed:
style dashboard components

tessalt committed:
make dashboard site-aware

tessalt committed:
change new commenters to new signups

tessalt committed:
generate new signups by day chart

tessalt committed:
update status pie chart

tessalt committed:
update status pie chart

tessalt committed:
allow site selection in dashboard

tessalt committed:
cache daily signups

tessalt committed:
count user bans in redis

tessalt committed:
clean up route and method names

tessalt committed:
clean up comment statuses chart and change from pie to bar

tessalt committed:
fix package lock

wyattjoh commented on package.json:
```diff
@@ -133,6 +134,8 @@
     "querystringify": "^2.1.1",
     "source-map-support": "^0.5.16",
     "stack-utils": "^2.0.1",
+    "react-helmet": "^5.2.1",
```
I'm guessing this was added by mistake?

wyattjoh commented on src/core/client/admin/routes/Dashboard/CommentActivity.tsx:
```diff
@@ -0,0 +1,95 @@
+          staffCount: comment.byAuthorRole.staff.count,
+          timestamp: new Date(comment.hour).getTime(),
+        };
+      });
```
As this is just from a REST route, why not perform this operation on the server?

wyattjoh commented on src/core/client/admin/routes/Dashboard/CommentStatuses.tsx:
```diff
@@ -0,0 +1,86 @@
+  async (environment: Environment, variables: any, { rest }) => {
+    const url = `/dashboard/${
+      variables.siteID ? variables.siteID + "/" : ""
+    }community-health`;
```
Typically, when I see routes like this:

```
/dashboard/${siteID}/community-health

AND

/dashboard/community-health
```

I'd probably suggest instead using a query parameter to denote an optional filter:

```
/dashboard/community-health?siteID=${siteID}
```

wyattjoh commented on src/core/server/app/router/api/index.ts:
```diff
@@ -56,6 +61,32 @@ export function createAPIRouter(app: AppOptions, options: RouterOptions) {
+    loggedInMiddleware,
+    roleMiddleware([
+      GQLUSER_ROLE.ADMIN,
+      GQLUSER_ROLE.STAFF,
```
Technically, staff cannot login to the admin, so no need to include that role.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -3180,6 +3180,7 @@ type Query {
   queues returns information on queues used in Coral to manage
   """
   queues: Queues! @auth(roles: [ADMIN])
+  dailyCounts: DailyCounts!
```
Looks like a mistake

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -1049,3 +1049,42 @@ export async function retrieveRecentStatusCounts(
+
+  // Get all of the statuses.
+  const docs = await cursor.toArray();
+  /* eslint-disable-next-line */
```
Why the eslint disable here?

wyattjoh commented on src/core/server/services/comments/stats.ts:
```diff
@@ -0,0 +1,134 @@
+  now: Date
+) {
+  await updateHourlyCount(redis, tenantID, siteID, now, hourlyCommentCountKey);
+  await updateDailyCount(redis, tenantID, siteID, now, dailyCommentCountKey);
```
I'd wrap these into the same redis operation instead, using `MULTI` to prevent the back/forth with redis.

tessalt committed:
make daily counts time zone aware

tessalt committed:
count rejected comments

tessalt committed:
clean up today counts

tessalt committed:
count comments from site

tessalt committed:
store hourly keys in utc

tessalt committed:
show daily user signups

tessalt committed:
move siteID from params to query params

tessalt committed:
add average line for daily comments chart

tessalt committed:
make average comments count hourly

tessalt committed:
show percent values for rejected comments

tessalt committed:
style dashboard

tessalt committed:
simplify dashboard rest routes

tessalt committed:
style today totals and top stories

tessalt committed:
style signup and comment activity chart

tessalt committed:
add site selector to nav

wyattjoh commented on src/core/server/app/handlers/api/dashboard/index.ts:
```diff
@@ -0,0 +1,222 @@
+        mongo,
+        redis,
+        tenant.id
+      );
```
Wrap in `Promise.all`

wyattjoh commented on src/core/server/app/middleware/attachSite.ts:
```diff
@@ -0,0 +1,34 @@
+    if (sitesCount > 1) {
+      if (req.query.siteID) {
+        const site = await retrieveSite(
+          options.mongo,
```
If no siteID is specified, do not set the `req.site`.

wyattjoh commented on src/core/server/app/middleware/role.ts:
```diff
@@ -0,0 +1,21 @@
+      new UserForbiddenError(
+        "user does not have sufficient privileges",
+        req.originalUrl,
+        req.method
```
Suggestion, add:

```js
req.user ? req.user.id : undefined
```

wyattjoh commented on src/core/server/app/router/api/dashboard.ts:
```diff
@@ -0,0 +1,52 @@
+        max: 5,
+        prefix: "userID",
+        config,
+      });
```
Move outside of request callback.

wyattjoh commented on src/core/server/app/router/api/dashboard.ts:
```diff
@@ -0,0 +1,52 @@
+  const router = createAPIRouter();
+  if (process.env.NODE_ENV !== "development") {
+    router.use(requestLimiter(app));
+  }
```
Instead, use the `DISABLE_RATE_LIMITERS=TRUE` when in development.

wyattjoh commented on src/core/server/app/router/api/dashboard.ts:
```diff
@@ -0,0 +1,52 @@
+    router.use(requestLimiter(app));
+  }
+
+  router.get("/today", attachSite(app), todayCountersHandler(app));
```
Move `attachSite` to router `router.use(attachSite(app))`.

wyattjoh replied:
Suggest moving site query to handlers instead.

wyattjoh commented on src/core/server/app/router/api/index.ts:
```diff
@@ -56,6 +61,19 @@ export function createAPIRouter(app: AppOptions, options: RouterOptions) {
+      GQLUSER_ROLE.STAFF,
+      GQLUSER_ROLE.MODERATOR,
+    ]),
+    jsonMiddleware,
```
Remove body parser middleware, this is only get requests.

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -1049,3 +1049,70 @@ export async function retrieveRecentStatusCounts(
+    },
+    {
+      $group: {
+        _id: { year: "$year", month: "$month", day: "$day", hour: "$hour" },
```
Move contents of `$project` to `$group._id`.

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -1049,3 +1049,70 @@ export async function retrieveRecentStatusCounts(
+    },
+  ]);
+  return cursor.toArray();
+}
```
Remove (should not be used, count on site document).

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -1049,3 +1049,70 @@ export async function retrieveRecentStatusCounts(
+    _id: null;
+    avg: number;
+  }>(mongo).aggregate([
+    { $match: { tenantID, siteID } },
```
Add `TODO` for narrowing scope of comments by `createdAt` to not go over the entire collection.

wyattjoh commented on src/core/server/models/site/index.ts:
```diff
@@ -121,6 +124,10 @@ async function retrieveConnection(
   return resolveConnection(query, input, (_, index) => index + skip + 1);
 }
 
+export async function countTenantSites(mongo: Db, tenantID: string) {
+  return collection(mongo).find({ tenantID }).count();
+}
+
```
Remove.

wyattjoh commented on src/core/server/models/site/index.ts:
```diff
@@ -173,3 +180,25 @@ export const updateSiteCounts = (
+export function getSiteCommentCount(site: Site) {
+  if (!site) {
+    return 0;
+  }
```
Remove.

wyattjoh commented on src/core/server/models/site/index.ts:
```diff
@@ -173,3 +180,25 @@ export const updateSiteCounts = (
+) {
+  if (!site) {
+    return 0;
+  }
```
Remove.

wyattjoh commented on src/core/server/models/story/index.ts:
```diff
@@ -663,3 +665,71 @@ export async function setStoryMode(
+  // Get the key to store the timezone aware data.
+  const key = `tcst:${tenantID}:${siteID}:${zone}:${startOfDay.toSeconds()}`;
+  // Perform the union store of the data.
+  await redis.zunionstore(key, 24, ranges[0], ...ranges.slice(1));
```
Should add expiry to the new union store key, wrap the `zunionstore` in `MULTI`.

wyattjoh commented on src/core/server/models/user/user.ts:
```diff
@@ -682,6 +682,24 @@ export async function retrieveManyUsers(
   return ids.map((id) => users.find((user) => user.id === id) || null);
 }
 
+export async function countUsersForCreationDate(
```
```suggestion
export async function countUsersCreatedBetween(
```

wyattjoh commented on src/core/server/models/user/user.ts:
```diff
@@ -682,6 +682,24 @@ export async function retrieveManyUsers(
+  mongo: Db,
+  tenantID: string,
+  after: Date,
+  before: Date | null,
```
Remove `null` as option.

wyattjoh commented on src/core/server/models/user/user.ts:
```diff
@@ -682,6 +682,24 @@ export async function retrieveManyUsers(
+        $lt: before || now,
+      },
+    })
+    .count();
```
Replace with aggregation. Replace `$gt` with `$gte`.

wyattjoh commented on src/core/server/services/comments/stats.ts:
```diff
@@ -0,0 +1,212 @@
+export interface DailyCommentCounts {
+  total: number;
+  staff: number;
+}
```
Remove.

wyattjoh commented on src/core/server/services/comments/stats.ts:
```diff
@@ -0,0 +1,212 @@
+  if (
+    [GQLUSER_ROLE.ADMIN, GQLUSER_ROLE.MODERATOR, GQLUSER_ROLE.STAFF].includes(
+      user.role
+    )
```
Replace with:

```ts
roleIsStaff(user.role)
```

wyattjoh commented on src/core/server/services/users/stats.ts:
```diff
@@ -0,0 +1,114 @@
+    tenant.auth.integrations.sso.allowRegistration
+  ) {
+    throw new Error("can't count users");
+  }
```
Remove.

wyattjoh commented on src/core/server/stacks/rejectComment.ts:
```diff
@@ -37,6 +41,12 @@ const rejectComment = async (
+  if (!comment) {
+    throw new CommentNotFoundError(commentID);
+  }
+
```
Instead, to get the `siteID`, do:

```ts
result.after.siteID
```

wyattjoh commented on src/core/server/app/router/api/index.ts:
```diff
@@ -56,6 +61,19 @@ export function createAPIRouter(app: AppOptions, options: RouterOptions) {
+    roleMiddleware([
+      GQLUSER_ROLE.ADMIN,
+      GQLUSER_ROLE.STAFF,
+      GQLUSER_ROLE.MODERATOR,
```
Replace with `STAFF_ROLES` from `src/core/server/models/user/constants.ts`

wyattjoh committed:
feat: redis/mongo refactor

wyattjoh committed:
fix: some small tweaks

- Added comments to magic numbers
- Added errors for missing input
- Consolidated promise resolutions

tessalt committed:
Revert "add site selector to nav"

This reverts commit 1c2b1dee34fb2742b04932079fd45f7f3de98418.

tessalt committed:
show first site dashboard by default

tessalt committed:
style dashboard site switcher

tessalt committed:
updte snapshot

tessalt committed:
add dashboard 118n keys

tessalt committed:
udpate comment activiyt chart with legend and tooltip

tessalt committed:
implement refresh button and loading states

tessalt committed:
change colours of charts for today

tessalt committed:
adjust today values spacing

tessalt committed:
remove unused package

tessalt committed:
don't pass redis into oauth strat

tessalt committed:
remove unused package

tessalt committed:
Revert "remove unused package"

This reverts commit 5b7c83a072604810ce7097865655e9ef8114d9e0.

nick-funk commented:
<img width="732" alt="image" src="https://user-images.githubusercontent.com/5751504/84049669-f37df580-a969-11ea-90f8-46a10649a1b3.png">

Small visual bug, the backgrounds on the dashboard icon blocks seem to be squishing and expanding as you resize the window width.

nick-funk commented on src/core/client/admin/routes/Dashboard/createDashboardFetch.ts:
```diff
@@ -0,0 +1,26 @@
+      params.set("tz", timeZone);
+
+      // // FIXME: remove, date forced for development
+      // params.set("date", "2020-05-05T12:30:00.000Z");
```
was this supposed to be removed?

wyattjoh replied:
Nope! This should be ok. It _could_ be changed to a `NOTE:` but it's not required.

nick-funk commented on src/core/client/admin/routes/Dashboard/sections/ChartColors.ts:
```diff
@@ -0,0 +1,8 @@
+export const CHART_COLOR_PRIMARY = "#419EA7";
+export const CHART_COLOR_PRIMARY_LIGHT = "#59C3C3";
+export const CHART_COLOR_SECONDARY = "#F77160";
+export const CHART_COLOR_MONO_500 = "#353F44";
+export const CHART_COLOR_MONO_100 = "#65696B";
+export const CHART_COLOR_GREY_200 = "#EAEFF0";
+export const CHART_COLOR_PRIMARY_PALE = "#9FECDF";
+export const CHART_COLOR_PRIMARY_DARK = "#2C7B8C";
```
+1, thank you for making things like this readily accessible and discoverable!

tessalt committed:
fix merge conflicts

tessalt committed:
Fix icons bug on smaller screens

tessalt committed:
resolve merge conflict
